numpyro/util.py

```diff
@@ -270,14 +275,369 @@ def update_fn(t, accept_prob, z_flat, state):
+    if half_tree.turning | half_tree.diverging:
+        return half_tree
+    else:
+        return _double_tree(half_tree, vv_update, kinetic_fn, _uniform_transition_prob,
+                            inverse_mass_matrix, step_size, going_right, doubling_key,
+                            energy_current, slice_exp_term, max_sliced_energy,
+                            use_multinomial_sampling, depth, iterative_build=False)
+
+
+@jit
+def _get_leaf(tree, going_right):
+    return lax.cond(going_right,
+                    tree,
+                    lambda tree: (tree.z_right, tree.r_right, tree.z_right_grad),
+                    tree,
+                    lambda tree: (tree.z_left, tree.r_left, tree.z_left_grad))
+
+
+def _double_tree(current_tree, vv_update, kinetic_fn, get_transition_prob,
+                 inverse_mass_matrix, step_size, going_right, rng,
+                 energy_current, slice_exp_term, max_sliced_energy,
+                 use_multinomial_sampling, max_tree_depth, iterative_build):
+    key, transition_key = random.split(rng)
+    # If we are going to the right, start from the right leaf of the current tree.
+    z, r, z_grad = _get_leaf(current_tree, going_right)
+
+    # Then build a new tree.
+    if iterative_build:
+        new_tree = _iterative_build_subtree(current_tree.depth, vv_update, kinetic_fn,
+                                            z, r, z_grad, inverse_mass_matrix, step_size,
+                                            going_right, key,
+                                            energy_current, slice_exp_term, max_sliced_energy,
+                                            use_multinomial_sampling, max_tree_depth)
+    else:
+        new_tree = _build_subtree(current_tree.depth, vv_update, kinetic_fn, z, r, z_grad,
+                                  inverse_mass_matrix, step_size, going_right, key,
+                                  energy_current, slice_exp_term, max_sliced_energy,
+                                  use_multinomial_sampling)
+    return _combine_tree(current_tree, new_tree, inverse_mass_matrix, going_right, transition_key,
+                         use_multinomial_sampling, get_transition_prob, iterative_build=False)
+
+
+def _leaf_idx_to_ckpt_idx(n):
+    # computes the number of non-zero bits except the last bit
+    # e.g. 6 -> 2, 7 -> 2, 13 -> 2
+    _, idx_max = lax.while_loop(lambda nc: nc[0] > 0,
+                                lambda nc: (nc[0] >> 1, nc[1] + (nc[0] & 1)),
+                                (n >> 1, 0))
+    # computes the number of last non-zero bits
+    # e.g. 6 -> 0, 7 -> 3, 13 -> 1
+    _, num_subtrees = lax.while_loop(lambda nc: (nc[0] & 1) != 0,
+                                     lambda nc: (nc[0] >> 1, nc[1] + 1),
+                                     (n, 0))
+    idx_min = idx_max - num_subtrees + 1
+    return idx_min, idx_max
+
+
+def _is_iterative_turning(leaf_idx, inverse_mass_matrix, r, r_sum, r_ckpts, r_sum_ckpts):
```

Comment:
If I understand correctly, this is doing two things:
 - updating checkpoints when `leaf_idx` is even.
 - checking for turning condition. I think we should only have to do this when `leaf_idx` is odd, right?

If so, it might be better to separate it out into two conditions `_update_checkpoints` and `_is_iterative_turning` which can be exercised by `lax.cond` based on whether the index is even or odd. Otherwise, we should probably rename this function appropriately.

Reply:
You are correct. When `leaf_idx` is odd, `ckpt_min > ckpt_max` so turning condition won't be updated.

The only thing which blocks me from separating is we have to recalculate `idx_min`, `idx_max`. I'll we move this calculation to the iterative body function and add these idxs to args of the methods `_update_checkpoints` and `_is_iterative_turning`.